browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -91,93 +89,77 @@ bool GetPolyfillForAdBlock(bool allow_brave_shields, bool allow_ads,
   return false;
 }
 
-void OnBeforeURLRequest_AdBlockTPCheckWork(
-    net::URLRequest* request,
-    GURL* new_url,
-    std::shared_ptr<BraveRequestInfo> ctx) {
-  // Proper content settings can't be looked up, so do nothing.
-  GURL url(request->url());
-  GURL tab_origin = request->site_for_cookies().GetOrigin();
-  if (tab_origin.is_empty() || !tab_origin.has_host() || url.is_empty()) {
+void OnBeforeURLRequestAdBlockTPOnTaskRunner(
+    GURL* new_url, std::shared_ptr<BraveRequestInfo> ctx) {
+  // If the following info isn't available, then proper content settings can't
+  // be looked up, so do nothing.
+  if (ctx->tab_origin.is_empty() || !ctx->tab_origin.has_host() ||
+      ctx->request_url.is_empty()) {
     return;
   }
   DCHECK(ctx->request_identifier != 0);
   if (!g_brave_browser_process->tracking_protection_service()->
-      ShouldStartRequest(url, ctx->resource_type, tab_origin.host())) {
+      ShouldStartRequest(ctx->request_url, ctx->resource_type, ctx->tab_origin.host())) {
     ctx->new_url_spec = GetBlankDataURLForResourceType(ctx->resource_type).spec();
   } else if (!g_brave_browser_process->ad_block_service()->ShouldStartRequest(
-           url, ctx->resource_type, tab_origin.host()) ||
+           ctx->request_url, ctx->resource_type, ctx->tab_origin.host()) ||
        !g_brave_browser_process->ad_block_regional_service()
-            ->ShouldStartRequest(url, ctx->resource_type,
-                                 tab_origin.host())) {
+            ->ShouldStartRequest(ctx->request_url, ctx->resource_type,
+                                 ctx->tab_origin.host())) {
     ctx->new_url_spec = GetBlankDataURLForResourceType(ctx->resource_type).spec();
   }
 }
 
-void OnBeforeURLRequest_OnBeforeURLRequest_AdBlockTPPostCheckWork(
-    net::URLRequest* request,
+void OnBeforeURLRequestDispatchOnIOThread(
     GURL* new_url,
     const ResponseCallback& next_callback,
     std::shared_ptr<BraveRequestInfo> ctx) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
   if (!ctx->new_url_spec.empty() &&
-    ctx->new_url_spec != request->url().spec()) {
+    ctx->new_url_spec != ctx->request_url.spec()) {
```

Comment:
follow up we should look at not passing `new_url` to the callbacks and instead having a const ctx.original_request_url and let the callbacks modify ctx.request_url
in the base delegate we can set new_url if request_url != original_request_url

Reply:
posted here for the follow-up:
https://github.com/brave/brave-browser/issues/1576